Introduce L7 load-balancing for requests to `kube-apiservers` (#11085)
gardener/gardener: docs/operations/kube_apiserver_loadbalancing.md

> # Kube API server load balancing

Reviewer: Please add a link to this to `docs/README.md`.

Load balancing for shoot and garden Kube API servers leverages the [Gardener Istio deployment](istio.md).

In default mode Istio performs L4 load balancing. Istio distributes connections to Kube API servers preferably in the
same zone ([Topology-Aware Traffic Routing](topology_aware_routing.md#topology-aware-services-in-the-seed-cluster)).
Requests are not taken into account for a load balancing decisions. TLS is terminated by Kube API server instances.  

There is a second mode which can be activated via `IstioTLSTermination` feature gate in `gardenlet` and `gardener-operator`.
This mode introduces L7 load balancing and distributes requests among the Kube API server instances.
It does not use [Topology-Aware Traffic Routing](topology_aware_routing.md) since it contradicts the load balancing goal.

> In this mode TLS is terminated in Istio. However, Istio creates an TLS to Kube API servers too so the traffic within the cluster remains encrypted.

Reviewer: ```suggestion
In this mode TLS is terminated in Istio. However, Istio creates a TLS connection to Kube API servers, too, so the traffic within the cluster remains encrypted.
```

This document is focused on the second mode.

## How it works

L7 load balancing currently works for the externally resolvable Kube API server endpoints only. Connections which use

> `apiserver-proxy` like `kubernetes.svc.cluster.local` still use L4 load balancing.

Reviewer: Nit: I prefer the full name.

```suggestion
`apiserver-proxy` like `kubernetes.default.svc.cluster.local` still use L4 load balancing.
```

> Clients might authenticate at Kube API server using client certificates or tokens. In the first case Istio ingress gateway

Reviewer: What about the case when clients do not authenticate? What happens then? I suppose we disable unauthenticated access per default, but client can still try, though.

Author: I added something about unauthenticated access.

mutual TLS authentication. It is optional because in case of the authentication via token, the token is validated by
Kube API server. For TLS termination and client certificate validation Kube API server certificate
(including the private key) and the client CA bundle must be synchronized into the namespace where istio ingress gateway
is running.

While token authentication is still done by Kube API server itself, it requires [dedicated request headers](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy)
if an authentication proxy is used. Based on these headers Kube API server identifies the user and group of a request.
They are added with a lua script running in an envoy filter in istio ingress gateway. Kube API server trusts the headers
set by the authentication proxy so it requires an mTLS connection to the proxy. Istio uses the front proxy CA
(including private key) and the server CA bundle to establish the mTLS connection. These secrets again must be synchronized
into the namespace where istio ingress gateway is running. There is an own istio destination rule for the mTLS connection.

> The istio virtual service points to an istio destination rule which does not use mTLS so that Kube API server authenticates

Reviewer: Not sure I would write it like that. As far as I remember, virtual services do not point to a destination rule, but there may be a destination rule for a virtual service.

Author: Updated 🙂

There is a second mode which can be activated via `IstioTLSTermination` feature gate in `gardenlet` and `gardener-operator`.
This mode introduces L7 load balancing and distributes requests among the Kube API server instances.
It does not use [Topology-Aware Traffic Routing](topology_aware_routing.md) since it contradicts the load balancing goal.
In this mode TLS is terminated in Istio. However, Istio creates a TLS connection to Kube API servers, too, so the
traffic within the cluster remains encrypted.

This document is focused on the second mode.

The minimum version of Kube API server (for virtual garden and shoot) is v1.31.0. Starting with this version the Kube API
server uses the WebSocket protocol instead of SPDY for streaming APIs ([ref](https://kubernetes.io/blog/2024/08/20/websockets-transition/)).
Envoy cannot upgrade the connection from HTTP to SPDY ([ref](https://github.com/envoyproxy/envoy/issues/36469)).  
Thus, for shoots and gardens using lower versions L7 load balancing remains deactivated even when the feature gate is enabled.  
On eligible shoots L7 load balancing is activated by default. However, it can be deactivated by annotating the shoot with

> `shoot.gardener.cloud/disable-istio-tls-termination: "true"`.

Reviewer: Should we mention that this also depends on the client? If I remember correctly it depends on the user updating their `kubectl` client. As it is valid practice to use a `v1.30` client with kubernetes `v1.31`, I think this should be documented as well.
WDYT?

## How it works

L7 load balancing currently works for the externally resolvable Kube API server endpoints and for connections which use

> `apiserver-proxy` like endpoint `kubernetes.default.svc.cluster.local`.

Reviewer: ```suggestion
L7 load balancing works for the externally resolvable Kube API server endpoints and for connections which use
`apiserver-proxy` like endpoint `kubernetes.default.svc.cluster.local`.
```

> Clients might authenticate at Kube API server using client certificates, tokens or might connect unauthenticated. In the

Reviewer: Just wondering: Is there a valid use case for unauthenticated connections in the Gardener ecosystem? If not, we could also handle it directly at istio to reduce `kube-apiserver` load. However, this is rather a performance optimisation.
WDYT?

Author: We allow users to `enableAnonymousAuthentication` ([ref](https://github.com/gardener/gardener/blob/master/docs/api-reference/core.md#kubeapiserverconfig)) for their shoots so there might be people using it.
At the moment there is only one envoy script for the public istio listener which is not aware of this setting. It would be another layer of complexity to change this so I think it would be more pain than gain 😅
